pennylane_lightning/src/algorithms/Observables.hpp

```diff
@@ -0,0 +1,419 @@
+            }
+        }
+        return true;
+    }
+
+  public:
+    /**
+     * @brief Create a tensor product of observables
+     *
+     * @param arg Arguments perfect forwarded to vector of observables.
+     */
+    template <typename... Ts>
+    explicit TensorProdObs(Ts &&...arg) : obs_{std::forward<Ts>(arg)...} {
+        std::unordered_set<size_t> wires;
+
+        for (const auto &ob : obs_) {
+            const auto ob_wires = ob->getWires();
+            for (const auto wire : ob_wires) {
+                if (wires.contains(wire)) {
+                    PL_ABORT("All wires in observables must be disjoint.");
+                }
+                wires.insert(wire);
+            }
+        }
+        all_wires_ = std::vector<size_t>(wires.begin(), wires.end());
+        std::sort(all_wires_.begin(), all_wires_.end());
+    }
+
+    /**
+     * @brief Get the number of operations in observable.
+     *
+     * @return size_t
+     */
+    [[nodiscard]] auto getSize() const -> size_t { return obs_.size(); }
+
+    /**
+     * @brief Get the wires for each observable operation.
+     *
+     * @return const std::vector<std::vector<size_t>>&
+     */
+    [[nodiscard]] auto getWires() const -> std::vector<size_t> final {
+        return all_wires_;
+    }
+
+    void applyInPlace(StateVectorManagedCPU<T> &sv) const final {
+        for (const auto &ob : obs_) {
+            ob->applyInPlace(sv);
+        }
+    }
+
+    [[nodiscard]] auto getObsName() const -> std::string final {
+        using Util::operator<<;
+        std::ostringstream obs_stream;
+        obs_stream << "Observable: { 'name' : ";
+        const auto obs_size = obs_.size();
+        for (size_t idx = 0; idx < obs_size; idx++) {
+            obs_stream << obs_[idx]->getObsName();
+            if (idx != obs_size - 1) {
+                obs_stream << " @ ";
+            }
+        }
+        obs_stream << ", 'wires' : " << getWires() << " }";
+        return obs_stream.str();
+    }
+};
+/// @cond DEV
+namespace detail {
+// Default implementation
+template <class T, bool use_openmp> struct HamiltonianApplyInPlace {
+    static void run(const std::vector<T> &coeffs,
+                    const std::vector<std::shared_ptr<Observable<T>>> &terms,
+                    StateVectorManagedCPU<T> &sv) {
+        auto allocator = sv.allocator();
+        std::vector<std::complex<T>, decltype(allocator)> res(
+            sv.getLength(), std::complex<T>{0.0, 0.0}, allocator);
+        for (size_t term_idx = 0; term_idx < coeffs.size(); term_idx++) {
+            StateVectorManagedCPU<T> tmp(sv);
+            terms[term_idx]->applyInPlace(tmp);
+            Util::scaleAndAdd(tmp.getLength(),
+                              std::complex<T>{coeffs[term_idx], 0.0},
+                              tmp.getData(), res.data());
+        }
+        sv.updateData(res);
+    }
+};
+#if defined(_OPENMP)
+template <class T> struct HamiltonianApplyInPlace<T, true> {
```

Comment:
I assume this will be run if available?

Reply:
I think so. Let me check.

Reply:
Yes this version is used when OpenMP is enabled.